lxd/device/device_utils_unix.go

```diff
@@ -176,13 +186,19 @@ func UnixDeviceCreate(s *state.State, idmapSet *idmap.IdmapSet, devicesPath stri
 		if err != nil {
 			return nil, fmt.Errorf("Invalid uid %s in device %s", m["uid"], srcPath)
 		}
+	} else {
+		_, uid, _ := unixDeviceOwnership(srcPath)
+		d.UID = int(uid)
```

Comment:
Here we are coercing a `uint32` into a `int`, right?

Reply:
Yes 👍

Reply:
I know we do this elsewhere but it doesn't feel like the right thing to do :/

Reply:
I don't think there is any potential for overflow here, but we could consider modifying the structure:

```Go
// UnixDevice contains information about a created UNIX device.
type UnixDevice struct {
	HostPath     string      // Absolute path to the device on the host.
	RelativePath string      // Relative path where the device will be mounted inside instance.
	Type         string      // Type of device; c (for char) or b for (block).
	Major        uint32      // Major number.
	Minor        uint32      // Minor number.
	Mode         os.FileMode // File mode.
	UID          int         // Owner UID.
	GID          int         // Owner GID.
}
```

I don't see any reason why the `UID` and `GID` fields can't be of type `uint32`. What do you think?

Reply:
Or, we can add some overflow validation to `unixDeviceOwnership` and return the gid and uid with type `int`.

Reply:
But since `unixDeviceOwnership` returns the gid and uid with type `uint32`, we aren't at risk for any overflow by using the `int()` cast function.

Reply:
As pointed out by Tom on another PR (https://github.com/canonical/lxd/pull/14327#discussion_r1829047630), `int()` on 32-bit platforms is only 32 bit long. This means it can overflow there, isn't it?

Reply:
In this case, it can't overflow because the gid and uid returned by `unixDeviceOwnership` is guaranteed to be of type `uint32`. I'd prefer either keeping the `int()` cast as is or adding an `int()` cast to the `unixDeviceOwnership` return values rather than modifying the `UnixDevice` structure.

Reply:
So we discussed over IM and got to the conclusion that despite `uint32()` risking to overflow once cast to `int()`, it was harmless as `int()` always return `-1` when wrapping around and @kadinsayani found it:

```
// Chown changes the numeric uid and gid of the named file.
// If the file is a symbolic link, it changes the uid and gid of the link's target.
// A uid or gid of -1 means to not change that value.
// If there is an error, it will be of type [*PathError].
//
// On Windows or Plan 9, Chown always returns the [syscall.EWINDOWS] or
// EPLAN9 error, wrapped in *PathError.
func Chown(name string, uid, gid int) error {
    e := ignoringEINTR(func() error {
        return syscall.Chown(name, uid, gid)
    })
    if e != nil {
        return &PathError{Op: "chown", Path: name, Err: e}
    }
    return nil
}
```

Meaning that in case of overflow, the UID/GID will remain unchanged anyway.

Reply:
Thanks for summarizing our findings @simondeziel :)